core/src/main/java/io/questdb/cutlass/line/http/AbstractLineHttpSender.java

```diff
@@ -185,40 +215,118 @@ public static AbstractLineHttpSender createLineSender(
+                final int blacklistableCount = Math.min(64, hosts.size());
+                final long allBlacklistedMask = (blacklistableCount == 64) ? -1L : ((1L << blacklistableCount) - 1);
+                for (int i = 0; ; i++) {
+                    currentAddressIndex = i % hosts.size();
+                    if (currentAddressIndex < 64) {
+                        // we can blacklist only the first 64 hosts, that's fine, we don't expect more hosts anyway
+                        // and if there ever are more hosts then the side-effect is that we will retry on the same host
+                        // even after it returned a non-retryable error
+                        if (hostBlacklistBitmap == allBlacklistedMask) {
+                            // if all hosts are blacklisted, we can't retry
+                            break;
+                        }
+                        if ((hostBlacklistBitmap & (1L << currentAddressIndex)) != 0) {
+                            // this host is blacklisted, skip it
+                            continue;
                         }
                     }
-                } else if (Utf8s.equalsNcAscii("404", statusCode)) {
-                    // The client is unable to differentiate between a server shutdown and connecting to an older version.
-                    // So, the protocol is set to PROTOCOL_VERSION_V1 here for both scenarios.
-                    protocolVersion = PROTOCOL_VERSION_V1;
-                } else {
-                    StringSink sink = new StringSink();
-                    chunkedResponseToSink(response, sink);
-                    throw new LineSenderException(
-                            "Failed to detect server line protocol version [http-status=").put(statusCode)
-                            .put(", http-message=").put(sink)
-                            .put(']');
+
+                    final String host = hosts.getQuick(currentAddressIndex);
+                    final int port = ports.getQuick(currentAddressIndex);
+                    try {
+                        HttpClient.Request req = cli.newRequest(host, port).GET().url(clientConfiguration.getSettingsPath());
+                        HttpClient.ResponseHeaders response = req.send();
+                        response.await();
+                        DirectUtf8Sequence statusCode = response.getStatusCode();
+                        if (isSuccessResponse(statusCode)) {
+                            try (JsonSettingsParser parser = new JsonSettingsParser()) {
+                                parser.parse(response.getResponse());
+                                protocolVersion = parser.getDefaultProtocolVersion();
+                                if (parser.getMaxNameLen() != 0) {
+                                    maxNameLength = parser.getMaxNameLen();
+                                }
+                                if (parser.isAcceptingWrites()) {
```

Comment:
Shouldn't we move initialization of both `protocolVersion` and `maxNameLength` under this branch? Otherwise we may end up with protocol version that applies to a previously pinged server.